### PR TITLE
ddl: revert "ddl: remove unnecessary locking when adding an index"

### PR DESCRIPTION
**Pull request by wjhuang2016**
Reverts pingcap/tidb#29772

We found that PR #29772 may cause inconsistent index and data. So we need to revert it.

The problem is that. In a pessimistic transaction, it doesn't check index key.

Init x = 1，and add index
1. backfill txn t1 get start_ts = 10
2. t1 read x, x = 1
3. pessimistic txn t2 get start_ts = for_update_ts = 11
4. t2 update set x = 2
5. t2 lock row
6. t1 prewrite index 1->x, commit commit_ts = 20
7. t2 prewrite update x=2, delete index 1->x, put index 2->x，
8. t2（async commit）get commit ts = 12 then commit，but commit before t1
finally：index 1->x, 2->x；row x=2

image

```release-note
None
```

### CODE REVIEW

**Review by zimulala — APPROVED** [review]
LGTM

**Comment by bb7133** [comment]
Please add some more comments why we have to add it back now. @wjhuang2016

File: ddl/column.go [hunk 1]
```diff
@@ -1427,7 +1427,7 @@ func (w *updateColumnWorker) cleanRowMap() {
 	}
 }
 
-// BackfillDataInTxn will backfill the table record in a transaction.
+// BackfillDataInTxn will backfill the table record in a transaction, lock corresponding rowKey, if the value of rowKey is changed.
```
Comment (djshow832) [inline ddl/column.go:1430]:
```suggestion
// BackfillDataInTxn will backfill the table record in a transaction. A lock corresponds to a rowKey if the value of rowKey is changed.
```

File: ddl/index.go [hunk 2]
```diff
@@ -1293,8 +1293,8 @@ func (w *addIndexWorker) batchCheckUniqueKey(txn kv.Transaction, idxRecords []*i
 	return nil
 }
 
-// BackfillDataInTxn will backfill table index in a transaction. If the value of rowKey is changed, there must be some other transactions
-// update the row, result in write conflict, so the txn will rollback and retry.
+// BackfillDataInTxn will backfill table index in a transaction, lock corresponding rowKey, if the value of rowKey is changed,
+// indicate that index columns values may changed, index is not allowed to be added, so the txn will rollback and retry.
```
Comment (djshow832) [inline ddl/index.go:1297]:
```suggestion
// BackfillDataInTxn will backfill table index in a transaction. A lock corresponds to a rowKey if the value of rowKey is changed,
// Note that index columns values may change, and an index is not allowed to be added, so the txn will rollback and retry.
```

**Review by djshow832 — COMMENTED** [review]
Although the comments were not written by you, it's really hard to understand for me :-(

**Review by djshow832 — APPROVED** [review]

**Pull request merged**